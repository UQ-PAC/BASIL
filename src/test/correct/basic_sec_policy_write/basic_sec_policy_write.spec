Globals:
x: int
z: int

L: x -> z == 0bv32, z -> true
Rely: old(z) == z
Guarantee: old(z) != 0bv32 ==> z != 0bv32

Subroutine: main
Requires: Gamma_R0 == false
