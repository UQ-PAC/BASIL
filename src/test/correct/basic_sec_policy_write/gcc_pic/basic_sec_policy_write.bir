000005cc: program
0000059a: sub __cxa_finalize(__cxa_finalize_result)
000005cd: __cxa_finalize_result :: out u32 = low:32[R0]

000003af:
000004a7: R16 := 0x20000
000004ae: R17 := mem[R16 + 8, el]:u64
000004b4: R16 := R16 + 8
000004b9: call R17 with noreturn

0000059b: sub __do_global_dtors_aux(__do_global_dtors_aux_result)
000005ce: __do_global_dtors_aux_result :: out u32 = low:32[R0]

0000029f:
000002a3: #3 := R31 - 0x20
000002a9: mem := mem with [#3, el]:u64 <- R29
000002af: mem := mem with [#3 + 8, el]:u64 <- R30
000002b3: R31 := #3
000002b9: R29 := R31
000002c1: mem := mem with [R31 + 0x10, el]:u64 <- R19
000002c6: R19 := 0x20000
000002cd: R0 := pad:64[mem[R19 + 0x30]]
000002d3: when 0:0[R0] goto %000002d1
000005c2: goto %00000378

00000378:
0000037b: R0 := 0x1F000
00000382: R0 := mem[R0 + 0xFB8, el]:u64
00000388: when R0 = 0 goto %00000386
000005c3: goto %0000039f

0000039f:
000003a2: R0 := 0x20000
000003a9: R0 := mem[R0 + 0x28, el]:u64
000003ae: R30 := 0x730
000003b1: call @__cxa_finalize with return %00000386

00000386:
0000038e: R30 := 0x734
00000390: call @deregister_tm_clones with return %00000392

00000392:
00000395: R0 := 1
0000039d: mem := mem with [R19 + 0x30] <- 7:0[R0]
000005c4: goto %000002d1

000002d1:
000002db: R19 := mem[R31 + 0x10, el]:u64
000002e2: R29 := mem[R31, el]:u64
000002e7: R30 := mem[R31 + 8, el]:u64
000002eb: R31 := R31 + 0x20
000002f0: call R30 with noreturn

0000059f: sub __libc_start_main(__libc_start_main_main, __libc_start_main_arg2, __libc_start_main_arg3, __libc_start_main_auxv, __libc_start_main_result)
000005cf: __libc_start_main_main :: in u64 = R0
000005d0: __libc_start_main_arg2 :: in u32 = low:32[R1]
000005d1: __libc_start_main_arg3 :: in out u64 = R2
000005d2: __libc_start_main_auxv :: in out u64 = R3
000005d3: __libc_start_main_result :: out u32 = low:32[R0]

000001f8:
00000491: R16 := 0x20000
00000498: R17 := mem[R16, el]:u64
0000049e: R16 := R16
000004a3: call R17 with noreturn

000005a0: sub _fini(_fini_result)
000005d4: _fini_result :: out u32 = low:32[R0]

00000020:
00000026: #0 := R31 - 0x10
0000002c: mem := mem with [#0, el]:u64 <- R29
00000032: mem := mem with [#0 + 8, el]:u64 <- R30
00000036: R31 := #0
0000003c: R29 := R31
00000043: R29 := mem[R31, el]:u64
00000048: R30 := mem[R31 + 8, el]:u64
0000004c: R31 := R31 + 0x10
00000051: call R30 with noreturn

000005a1: sub _init(_init_result)
000005d5: _init_result :: out u32 = low:32[R0]

00000501:
00000507: #5 := R31 - 0x10
0000050d: mem := mem with [#5, el]:u64 <- R29
00000513: mem := mem with [#5 + 8, el]:u64 <- R30
00000517: R31 := #5
0000051d: R29 := R31
00000522: R30 := 0x5C0
00000524: call @call_weak_fn with return %00000526

00000526:
0000052b: R29 := mem[R31, el]:u64
00000530: R30 := mem[R31 + 8, el]:u64
00000534: R31 := R31 + 0x10
00000539: call R30 with noreturn

000005a2: sub _start(_start_result)
000005d6: _start_result :: out u32 = low:32[R0]

000001b9:
000001be: R29 := 0
000001c3: R30 := 0
000001c9: R5 := R0
000001d0: R1 := mem[R31, el]:u64
000001d6: R2 := R31 + 8
000001dc: R6 := R31
000001e1: R0 := 0x1F000
000001e8: R0 := mem[R0 + 0xFD8, el]:u64
000001ed: R3 := 0
000001f2: R4 := 0
000001f7: R30 := 0x670
000001fa: call @__libc_start_main with return %000001fc

000001fc:
000001ff: R30 := 0x674
00000202: call @abort with return %000005c5

000005c5:
000005c6: call @call_weak_fn with noreturn

000005a5: sub abort()


00000200:
000004d3: R16 := 0x20000
000004da: R17 := mem[R16 + 0x18, el]:u64
000004e0: R16 := R16 + 0x18
000004e5: call R17 with noreturn

000005a6: sub call_weak_fn(call_weak_fn_result)
000005d7: call_weak_fn_result :: out u32 = low:32[R0]

00000204:
00000207: R0 := 0x1F000
0000020e: R0 := mem[R0 + 0xFD0, el]:u64
00000214: when R0 = 0 goto %00000212
000005c7: goto %000003ef

00000212:
0000021a: call R30 with noreturn

000003ef:
000003f2: goto @__gmon_start__

000003f0:
000004bd: R16 := 0x20000
000004c4: R17 := mem[R16 + 0x10, el]:u64
000004ca: R16 := R16 + 0x10
000004cf: call R17 with noreturn

000005a8: sub deregister_tm_clones(deregister_tm_clones_result)
000005d8: deregister_tm_clones_result :: out u32 = low:32[R0]

00000220:
00000223: R0 := 0x20000
00000229: R0 := R0 + 0x30
0000022e: R1 := 0x20000
00000234: R1 := R1 + 0x30
0000023a: #1 := ~R0
0000023f: #2 := R1 + ~R0
00000245: VF := extend:65[#2 + 1] <> extend:65[R1] + extend:65[#1] + 1
0000024b: CF := pad:65[#2 + 1] <> pad:65[R1] + pad:65[#1] + 1
0000024f: ZF := #2 + 1 = 0
00000253: NF := 63:63[#2 + 1]
00000259: when ZF goto %00000257
000005c8: goto %000003d1

000003d1:
000003d4: R1 := 0x1F000
000003db: R1 := mem[R1 + 0xFB0, el]:u64
000003e0: when R1 = 0 goto %00000257
000005c9: goto %000003e4

00000257:
0000025f: call R30 with noreturn

000003e4:
000003e8: R16 := R1
000003ed: call R16 with noreturn

000005ab: sub frame_dummy(frame_dummy_result)
000005d9: frame_dummy_result :: out u32 = low:32[R0]

000002f6:
000002f8: call @register_tm_clones with noreturn

000005ac: sub main(main_argc, main_argv, main_result)
000005da: main_argc :: in u32 = low:32[R0]
000005db: main_argv :: in out u64 = R1
000005dc: main_result :: out u32 = low:32[R0]

000002fa:
000002fe: R31 := R31 - 0x10
00000306: mem := mem with [R31 + 0xC, el]:u32 <- 31:0[R0]
0000030b: R0 := 0x1F000
00000312: R0 := mem[R0 + 0xFC0, el]:u64
00000317: R1 := 1
0000031f: mem := mem with [R0, el]:u32 <- 31:0[R1]
00000324: R0 := 0x1F000
0000032b: R0 := mem[R0 + 0xFC8, el]:u64
00000332: R1 := pad:64[mem[R31 + 0xC, el]:u32]
0000033a: mem := mem with [R0, el]:u32 <- 31:0[R1]
0000033f: R0 := 0x1F000
00000346: R0 := mem[R0 + 0xFC8, el]:u64
0000034d: mem := mem with [R0, el]:u32 <- 0
00000352: R0 := 0x1F000
00000359: R0 := mem[R0 + 0xFC0, el]:u64
0000035e: R1 := 2
00000366: mem := mem with [R0, el]:u32 <- 31:0[R1]
0000036b: R0 := 0
00000371: R31 := R31 + 0x10
00000376: call R30 with noreturn

000005ad: sub register_tm_clones(register_tm_clones_result)
000005dd: register_tm_clones_result :: out u32 = low:32[R0]

00000261:
00000264: R0 := 0x20000
0000026a: R0 := R0 + 0x30
0000026f: R1 := 0x20000
00000275: R1 := R1 + 0x30
0000027c: R1 := R1 + ~R0 + 1
00000282: R2 := 0.63:63[R1]
00000289: R1 := R2 + (R1 ~>> 3)
0000028f: R1 := extend:64[63:1[R1]]
00000295: when R1 = 0 goto %00000293
000005ca: goto %000003b3

000003b3:
000003b6: R2 := 0x1F000
000003bd: R2 := mem[R2 + 0xFE0, el]:u64
000003c2: when R2 = 0 goto %00000293
000005cb: goto %000003c6

00000293:
0000029b: call R30 with noreturn

000003c6:
000003ca: R16 := R2
000003cf: call R16 with noreturn
