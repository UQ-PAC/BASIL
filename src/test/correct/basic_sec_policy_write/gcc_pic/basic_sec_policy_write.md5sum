0d579d5f95d5b9ada274b8d277a5274d  correct/basic_sec_policy_write/gcc_pic/a.out
bdb4e5a39334379f2a320c6e6cbbcd7c  correct/basic_sec_policy_write/gcc_pic/basic_sec_policy_write.adt
e3593af37d2ad4872f495f2e14a9009d  correct/basic_sec_policy_write/gcc_pic/basic_sec_policy_write.bir
13288c272806c0a724285c7a07336f2f  correct/basic_sec_policy_write/gcc_pic/basic_sec_policy_write.relf
084a854a83e4d403a6182e1c435dee72  correct/basic_sec_policy_write/gcc_pic/basic_sec_policy_write.gts
