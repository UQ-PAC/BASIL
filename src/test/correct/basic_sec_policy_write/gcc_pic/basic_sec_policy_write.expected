var Gamma_R0: bool;
var Gamma_R1: bool;
var Gamma_R31: bool;
var Gamma_mem: [bv64]bool;
var Gamma_stack: [bv64]bool;
var R0: bv64;
var R1: bv64;
var R31: bv64;
var mem: [bv64]bv8;
var stack: [bv64]bv8;
const $x_addr: bv64;
axiom ($x_addr == 69652bv64);
const $z_addr: bv64;
axiom ($z_addr == 69656bv64);
function L(memory: [bv64]bv8, index: bv64) returns (bool) {
  (if (index == $z_addr) then true else (if (index == $x_addr) then (memory_load32_le(memory, $z_addr) == 0bv32) else false))
}

function {:bvbuiltin "bvadd"} bvadd64(bv64, bv64) returns (bv64);
function gamma_load32(gammaMap: [bv64]bool, index: bv64) returns (bool) {
  (gammaMap[bvadd64(index, 3bv64)] && (gammaMap[bvadd64(index, 2bv64)] && (gammaMap[bvadd64(index, 1bv64)] && gammaMap[index])))
}

function gamma_load64(gammaMap: [bv64]bool, index: bv64) returns (bool) {
  (gammaMap[bvadd64(index, 7bv64)] && (gammaMap[bvadd64(index, 6bv64)] && (gammaMap[bvadd64(index, 5bv64)] && (gammaMap[bvadd64(index, 4bv64)] && (gammaMap[bvadd64(index, 3bv64)] && (gammaMap[bvadd64(index, 2bv64)] && (gammaMap[bvadd64(index, 1bv64)] && gammaMap[index])))))))
}

function gamma_store32(gammaMap: [bv64]bool, index: bv64, value: bool) returns ([bv64]bool) {
  gammaMap[index := value][bvadd64(index, 1bv64) := value][bvadd64(index, 2bv64) := value][bvadd64(index, 3bv64) := value]
}

function memory_load32_le(memory: [bv64]bv8, index: bv64) returns (bv32) {
  (memory[bvadd64(index, 3bv64)] ++ (memory[bvadd64(index, 2bv64)] ++ (memory[bvadd64(index, 1bv64)] ++ memory[index])))
}

function memory_load64_le(memory: [bv64]bv8, index: bv64) returns (bv64) {
  (memory[bvadd64(index, 7bv64)] ++ (memory[bvadd64(index, 6bv64)] ++ (memory[bvadd64(index, 5bv64)] ++ (memory[bvadd64(index, 4bv64)] ++ (memory[bvadd64(index, 3bv64)] ++ (memory[bvadd64(index, 2bv64)] ++ (memory[bvadd64(index, 1bv64)] ++ memory[index])))))))
}

function memory_load8_le(memory: [bv64]bv8, index: bv64) returns (bv8) {
  memory[index]
}

function memory_store32_le(memory: [bv64]bv8, index: bv64, value: bv32) returns ([bv64]bv8) {
  memory[index := value[8:0]][bvadd64(index, 1bv64) := value[16:8]][bvadd64(index, 2bv64) := value[24:16]][bvadd64(index, 3bv64) := value[32:24]]
}

function {:bvbuiltin "zero_extend 32"} zero_extend32_32(bv32) returns (bv64);
procedure rely();
  modifies Gamma_mem, mem;
  ensures (forall i: bv64 :: (((mem[i] == old(mem[i])) ==> (Gamma_mem[i] == old(Gamma_mem[i])))));
  ensures (old(memory_load32_le(mem, $z_addr)) == memory_load32_le(mem, $z_addr));
  free ensures (memory_load8_le(mem, 1976bv64) == 1bv8);
  free ensures (memory_load8_le(mem, 1977bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 1978bv64) == 2bv8);
  free ensures (memory_load8_le(mem, 1979bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69640bv64) == 8bv8);
  free ensures (memory_load8_le(mem, 69641bv64) == 16bv8);
  free ensures (memory_load8_le(mem, 69642bv64) == 1bv8);
  free ensures (memory_load8_le(mem, 69643bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69644bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69645bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69646bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69647bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69616bv64) == 84bv8);
  free ensures (memory_load8_le(mem, 69617bv64) == 7bv8);
  free ensures (memory_load8_le(mem, 69618bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69619bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69620bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69621bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69622bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69623bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69600bv64) == 20bv8);
  free ensures (memory_load8_le(mem, 69601bv64) == 16bv8);
  free ensures (memory_load8_le(mem, 69602bv64) == 1bv8);
  free ensures (memory_load8_le(mem, 69603bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69604bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69605bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69606bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69607bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69008bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69009bv64) == 7bv8);
  free ensures (memory_load8_le(mem, 69010bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69011bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69012bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69013bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69014bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69015bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69592bv64) == 24bv8);
  free ensures (memory_load8_le(mem, 69593bv64) == 16bv8);
  free ensures (memory_load8_le(mem, 69594bv64) == 1bv8);
  free ensures (memory_load8_le(mem, 69595bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69596bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69597bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69598bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69599bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69000bv64) == 80bv8);
  free ensures (memory_load8_le(mem, 69001bv64) == 7bv8);
  free ensures (memory_load8_le(mem, 69002bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69003bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69004bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69005bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69006bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69007bv64) == 0bv8);

procedure rely_transitive()
  modifies Gamma_mem, mem;
  ensures (old(memory_load32_le(mem, $z_addr)) == memory_load32_le(mem, $z_addr));
{
  call rely();
  call rely();
}

procedure rely_reflexive()
{
  assert (memory_load32_le(mem, $z_addr) == memory_load32_le(mem, $z_addr));
}

procedure guarantee_reflexive()
  modifies Gamma_mem, mem;
{
  assert ((memory_load32_le(mem, $z_addr) != 0bv32) ==> (memory_load32_le(mem, $z_addr) != 0bv32));
}

procedure main()
  modifies Gamma_R0, Gamma_R1, Gamma_R31, Gamma_mem, Gamma_stack, R0, R1, R31, mem, stack;
  requires (Gamma_R0 == false);
  free requires (memory_load8_le(mem, 69632bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69633bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69634bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69635bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69636bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69637bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69638bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69639bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69640bv64) == 8bv8);
  free requires (memory_load8_le(mem, 69641bv64) == 16bv8);
  free requires (memory_load8_le(mem, 69642bv64) == 1bv8);
  free requires (memory_load8_le(mem, 69643bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69644bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69645bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69646bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69647bv64) == 0bv8);
  free requires (memory_load8_le(mem, 1976bv64) == 1bv8);
  free requires (memory_load8_le(mem, 1977bv64) == 0bv8);
  free requires (memory_load8_le(mem, 1978bv64) == 2bv8);
  free requires (memory_load8_le(mem, 1979bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69640bv64) == 8bv8);
  free requires (memory_load8_le(mem, 69641bv64) == 16bv8);
  free requires (memory_load8_le(mem, 69642bv64) == 1bv8);
  free requires (memory_load8_le(mem, 69643bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69644bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69645bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69646bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69647bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69616bv64) == 84bv8);
  free requires (memory_load8_le(mem, 69617bv64) == 7bv8);
  free requires (memory_load8_le(mem, 69618bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69619bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69620bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69621bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69622bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69623bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69600bv64) == 20bv8);
  free requires (memory_load8_le(mem, 69601bv64) == 16bv8);
  free requires (memory_load8_le(mem, 69602bv64) == 1bv8);
  free requires (memory_load8_le(mem, 69603bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69604bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69605bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69606bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69607bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69008bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69009bv64) == 7bv8);
  free requires (memory_load8_le(mem, 69010bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69011bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69012bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69013bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69014bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69015bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69592bv64) == 24bv8);
  free requires (memory_load8_le(mem, 69593bv64) == 16bv8);
  free requires (memory_load8_le(mem, 69594bv64) == 1bv8);
  free requires (memory_load8_le(mem, 69595bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69596bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69597bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69598bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69599bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69000bv64) == 80bv8);
  free requires (memory_load8_le(mem, 69001bv64) == 7bv8);
  free requires (memory_load8_le(mem, 69002bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69003bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69004bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69005bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69006bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69007bv64) == 0bv8);
  free ensures (Gamma_R31 == old(Gamma_R31));
  free ensures (R31 == old(R31));
  free ensures (memory_load8_le(mem, 1976bv64) == 1bv8);
  free ensures (memory_load8_le(mem, 1977bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 1978bv64) == 2bv8);
  free ensures (memory_load8_le(mem, 1979bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69640bv64) == 8bv8);
  free ensures (memory_load8_le(mem, 69641bv64) == 16bv8);
  free ensures (memory_load8_le(mem, 69642bv64) == 1bv8);
  free ensures (memory_load8_le(mem, 69643bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69644bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69645bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69646bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69647bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69616bv64) == 84bv8);
  free ensures (memory_load8_le(mem, 69617bv64) == 7bv8);
  free ensures (memory_load8_le(mem, 69618bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69619bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69620bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69621bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69622bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69623bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69600bv64) == 20bv8);
  free ensures (memory_load8_le(mem, 69601bv64) == 16bv8);
  free ensures (memory_load8_le(mem, 69602bv64) == 1bv8);
  free ensures (memory_load8_le(mem, 69603bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69604bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69605bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69606bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69607bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69008bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69009bv64) == 7bv8);
  free ensures (memory_load8_le(mem, 69010bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69011bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69012bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69013bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69014bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69015bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69592bv64) == 24bv8);
  free ensures (memory_load8_le(mem, 69593bv64) == 16bv8);
  free ensures (memory_load8_le(mem, 69594bv64) == 1bv8);
  free ensures (memory_load8_le(mem, 69595bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69596bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69597bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69598bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69599bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69000bv64) == 80bv8);
  free ensures (memory_load8_le(mem, 69001bv64) == 7bv8);
  free ensures (memory_load8_le(mem, 69002bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69003bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69004bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69005bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69006bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69007bv64) == 0bv8);
{
  var Gamma_x_old: bool;
  var z_old: bv32;
  lmain:
    R31, Gamma_R31 := bvadd64(R31, 18446744073709551600bv64), Gamma_R31;
    stack, Gamma_stack := memory_store32_le(stack, bvadd64(R31, 12bv64), R0[32:0]), gamma_store32(Gamma_stack, bvadd64(R31, 12bv64), Gamma_R0);
    R0, Gamma_R0 := 65536bv64, true;
    call rely();
    R0, Gamma_R0 := memory_load64_le(mem, bvadd64(R0, 4056bv64)), (gamma_load64(Gamma_mem, bvadd64(R0, 4056bv64)) || L(mem, bvadd64(R0, 4056bv64)));
    R1, Gamma_R1 := 1bv64, true;
    call rely();
    assert (L(mem, R0) ==> Gamma_R1);
    z_old := memory_load32_le(mem, $z_addr);
    Gamma_x_old := (gamma_load32(Gamma_mem, $x_addr) || L(mem, $x_addr));
    mem, Gamma_mem := memory_store32_le(mem, R0, R1[32:0]), gamma_store32(Gamma_mem, R0, Gamma_R1);
    assert ((R0 == $z_addr) ==> (L(mem, $x_addr) ==> Gamma_x_old));
    assert ((z_old != 0bv32) ==> (memory_load32_le(mem, $z_addr) != 0bv32));
    R0, Gamma_R0 := 65536bv64, true;
    call rely();
    R0, Gamma_R0 := memory_load64_le(mem, bvadd64(R0, 4064bv64)), (gamma_load64(Gamma_mem, bvadd64(R0, 4064bv64)) || L(mem, bvadd64(R0, 4064bv64)));
    R1, Gamma_R1 := zero_extend32_32(memory_load32_le(stack, bvadd64(R31, 12bv64))), gamma_load32(Gamma_stack, bvadd64(R31, 12bv64));
    call rely();
    assert (L(mem, R0) ==> Gamma_R1);
    z_old := memory_load32_le(mem, $z_addr);
    Gamma_x_old := (gamma_load32(Gamma_mem, $x_addr) || L(mem, $x_addr));
    mem, Gamma_mem := memory_store32_le(mem, R0, R1[32:0]), gamma_store32(Gamma_mem, R0, Gamma_R1);
    assert ((R0 == $z_addr) ==> (L(mem, $x_addr) ==> Gamma_x_old));
    assert ((z_old != 0bv32) ==> (memory_load32_le(mem, $z_addr) != 0bv32));
    R0, Gamma_R0 := 65536bv64, true;
    call rely();
    R0, Gamma_R0 := memory_load64_le(mem, bvadd64(R0, 4064bv64)), (gamma_load64(Gamma_mem, bvadd64(R0, 4064bv64)) || L(mem, bvadd64(R0, 4064bv64)));
    call rely();
    assert (L(mem, R0) ==> true);
    z_old := memory_load32_le(mem, $z_addr);
    Gamma_x_old := (gamma_load32(Gamma_mem, $x_addr) || L(mem, $x_addr));
    mem, Gamma_mem := memory_store32_le(mem, R0, 0bv32), gamma_store32(Gamma_mem, R0, true);
    assert ((R0 == $z_addr) ==> (L(mem, $x_addr) ==> Gamma_x_old));
    assert ((z_old != 0bv32) ==> (memory_load32_le(mem, $z_addr) != 0bv32));
    R0, Gamma_R0 := 65536bv64, true;
    call rely();
    R0, Gamma_R0 := memory_load64_le(mem, bvadd64(R0, 4056bv64)), (gamma_load64(Gamma_mem, bvadd64(R0, 4056bv64)) || L(mem, bvadd64(R0, 4056bv64)));
    R1, Gamma_R1 := 2bv64, true;
    call rely();
    assert (L(mem, R0) ==> Gamma_R1);
    z_old := memory_load32_le(mem, $z_addr);
    Gamma_x_old := (gamma_load32(Gamma_mem, $x_addr) || L(mem, $x_addr));
    mem, Gamma_mem := memory_store32_le(mem, R0, R1[32:0]), gamma_store32(Gamma_mem, R0, Gamma_R1);
    assert ((R0 == $z_addr) ==> (L(mem, $x_addr) ==> Gamma_x_old));
    assert ((z_old != 0bv32) ==> (memory_load32_le(mem, $z_addr) != 0bv32));
    R0, Gamma_R0 := 0bv64, true;
    R31, Gamma_R31 := bvadd64(R31, 16bv64), Gamma_R31;
    return;
}
